workflowparsers/quantum_espresso_bands/parser.py

```diff
@@ -0,0 +1,570 @@
+                    quantities=[
+                        Quantity(
+                            'energy',
+                            r'e\(\s*\d+ -\s*\d+\) =\s*([\-\d\.]+)\s+eV',
+                            repeats=True,
+                            dtype=float,
+                        ),
+                        Quantity(
+                            'mult',
+                            r'eV\s*(\d+)\s*-->',
+                            repeats=True,
+                            dtype=int,
+                        ),
+                    ],
+                ),
+            ),
+            Quantity(
+                'failed_symmetry',
+                r'(zone border point and non-symmorphic group)',
+                repeats=True,
+                dtype=str,
+                default='',
+            ),
+        ]
+
+    @staticmethod
+    def scan_dir_for_files(directory: str, pattern: str = '*.out') -> list[str]:
+        """Find all files matching the pattern in the given directory."""
+        return glob.glob(os.path.join(directory, pattern))
+
+    @staticmethod
+    def points_to_segments(kpoints: list, symmetries: list) -> list[list[KPoint]]:
+        """Split the k-points by segment based on differing symmetry group."""
+
+        def shift_window(window: tuple, elem) -> tuple:
+            return window[1:] + (elem,)
+
+        previous_kpoint: Optional[KPoint] = None
+        symmetry_window: tuple[Optional[str], Optional[str], Optional[str]] = (
+            None,
+        ) * 3
+
+        segments: list[list[KPoint]] = [[]]  # segments are lists of k-point paths
+        for kpoint, symmetry in zip(kpoints, symmetries):
+            symmetry_window = shift_window(symmetry_window, symmetry)
+            # case enumeration for `symmetry_window`:
+            # 1. (None, None, None) -> not possible
+            # 2. (None, None, X) -> first step (add to initial bucket)
+            # 3. (X, None, None) -> not possible
+            # 4. (X, Y, None) -> end reached
+            # 5. (None, X, Y) -> add to initial bucket
+            # 6. (X, X, Y) -> add Y to latest bucket
+            # 7. (X, Y, Y) -> add Y to latest bucket
+            # 8. (X, Y, X) -> add Y and X a new, latest bucket
+            # 9. (X, Y, Z) -> add Y and Z a new, latest bucket
+            if (None not in symmetry_window) and all(
+                [symmetry_window[i] != symmetry_window[i + 1] for i in range(2)]
+            ):
+                segments.append([previous_kpoint])
+            segments[-1].append(kpoint)
+            previous_kpoint = kpoint
+        return segments
+
+    @staticmethod
+    def apply_multiplicity(
+        energies: list[list[float]], mults: list[list[int]]
+    ) -> list[list[float]]:
+        """Apply band multiplicity to energies."""
+        return [
+            [e for e, m in zip(energy, mult) for _ in range(m)]
+            for energy, mult in zip(energies, mults)
+        ]
+
+
+class QuantumEspressoBandsParser:
```

Comment:
add a get_mainfile_keys function, to scan the pwscf file, and if present should return an key for the child archive which contains the overall workflow

Reply:
> add a get_mainfile_keys function, to scan the pwscf file

This is currently exactly done by `_find_files`. Do you think that I should just change the name, the type signature, or something more?

> should return an key for the child archive which contains the overall workflow

How do you see child archives entering here? Both the information from PWSCF and BANDS have to be included at the same level.

Reply:
you define get_mainfile_keys to scan the pwscf files in the same directory or one above it, then if there is it will return the name of the file which you will later use to search the appropriate entry with nomad.search in the parse method. Additionally, you should also return another key for the overall workflow. Normally, we do the reverse and put the generation of the child_archives in the root parser in this case pwscf but yeah, i do not have a strong opinion on this tbh. Maybe put it then in the pwscf parser this child archive generation not in bands. if you do the second route, you should set pwscf parser.level to a higher value than bands so the bands entry is parsed and searchable when you connect it to the pwscf entry

Reply:
no there is no docs for this, maybe just refer to the MatchingParserInterface class. it should return a list of keys for your child archives, for the pwscf file, you should return the full path so you essentially match it the pwscf entry mainfile

Reply:
I'll make it return the key `bs_analysis_pwscf`

Reply:
do not anymore do this i will implement the linking of entries in the qe parser